core/src/parking_lot.rs

```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
+    /// Note: Call before the resource is released
+    #[inline]
+    pub fn release_resource(_key: usize) {
+        #[cfg(feature = "deadlock_detection")]
+        deadlock_impl::release_resource(_key);
+    }
+
+    /// TODO: doc
+    #[cfg(feature = "deadlock_detection")]
+    #[inline]
+    pub fn check_deadlock() -> bool {
+        deadlock_impl::check_deadlock()
+    }
+}
+
+#[cfg(feature = "deadlock_detection")]
+mod deadlock_impl {
+    use super::{get_hashtable, lock_bucket, ThreadData, get_thread_data, NUM_THREADS};
+    use std::sync::mpsc;
+    use std::sync::atomic::Ordering;
+    use std::collections::HashSet;
+    use backtrace::Backtrace;
+    use petgraph;
+    use petgraph::graphmap::DiGraphMap;
+
+    pub(super) unsafe fn on_unpark(td: &ThreadData) {
+        if td.deadlock_data.deadlocked.get() {
+            if let Some(sender) = (*td.deadlock_data.backtrace_sender.get()).take() {
+                let _ = sender.send((td.deadlock_data.thread_id, Backtrace::new()));
+            }
+            td.parker.prepare_park();
+            td.parker.park();
+            unreachable!("unparked deadlocked thread!");
+        }
+    }
+
     pub fn acquire_resource(key: usize) {
         unsafe {
             let mut thread_data = None;
             let thread_data = get_thread_data(&mut thread_data);
-            (*thread_data.resources.get()).push(key);
+            (*thread_data.deadlock_data.resources.get()).push(key);
         }
     }
 
-    /// Release a resource identified by key in the deadlock detector
-    /// Note: Call before the resource is released
     pub fn release_resource(key: usize) {
         unsafe {
             let mut thread_data = None;
             let thread_data = get_thread_data(&mut thread_data);
-            let resources = &mut (*thread_data.resources.get());
+            let resources = &mut (*thread_data.deadlock_data.resources.get());
             match resources.iter().rposition(|x| *x == key) {
                 Some(p) => resources.swap_remove(p),
-                None => {
-                    check_deadlock();
-                    panic!("key {} not found in thread resources", key);
-                },
+                None => panic!("key {} not found in thread resources", key),
             };
         }
     }
 
-    /// Returns true if a deadlock is detected
     pub fn check_deadlock() -> bool {
-        // basic algorithm that builds a directed graph of the threads and the resources
-        // then checks for the presence of cycles (deadlocks)
-
-        // fast pass
-        if check_wait_graph_cycle(false) {
-            // double check with a full hash table lock
-            check_wait_graph_cycle(true)
-        } else {
-            false
+        unsafe {
+            // fast pass
+            if check_wait_graph_fast() {
+                // double check
+                let deadlocks = check_wait_graph_slow();
+                // TODO: call a configurable callback
```

Comment:
Rather than using a callback, I would suggest returning a `struct` with all the backtraces of deadlocked threads.

Reply:
Indeed, it's also easier.